pkg/foundation/grpcutil/websocket.go

```diff
@@ -0,0 +1,132 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcutil
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"net/http"
+
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/gorilla/websocket"
+)
+
+type inMemoryResponseWriter struct {
+	io.Writer
+	header http.Header
+	closed chan bool
+}
+
+func newInMemoryResponseWriter(writer io.Writer) *inMemoryResponseWriter {
+	return &inMemoryResponseWriter{
+		Writer: writer,
+		header: http.Header{},
+		closed: make(chan bool, 1),
+	}
+}
+
+func (w *inMemoryResponseWriter) Write(b []byte) (int, error) {
+	return w.Writer.Write(b)
+}
+func (w *inMemoryResponseWriter) Header() http.Header {
+	return w.header
+}
+func (w *inMemoryResponseWriter) WriteHeader(int) {
+	// we don't have a use for the code
+}
+func (w *inMemoryResponseWriter) CloseNotify() <-chan bool {
+	return w.closed
+}
+func (w *inMemoryResponseWriter) Flush() {}
+
+// wsProxy is a proxy around a http.Handler which
+// redirects the response data from the http.Handler
+// to a WebSocket connection.
+type wsProxy struct {
+	handler  http.Handler
+	logger   log.CtxLogger
+	upgrader websocket.Upgrader
+}
+
+func newWebSocketProxy(handler http.Handler, logger log.CtxLogger) *wsProxy {
+	return &wsProxy{
+		handler: handler,
+		logger:  logger.WithComponent("grpcutil.websocket"),
```

Comment:
The reason why I put `websocket` is because wsProxy is an internal/unexported type and because it's more clear what is it about. But I don't have anything against wsProxy either.

Reply:
The function itself is also unexported, so I don't see the problem 🤷 The point of the logger component is to help us find where a log message originated from, so keeping that aligned with the struct name makes it easy to find it. All in all see no upside in using other names here.

Reply:
Gotcha, I was understanding `WithComponent` more as a logical component, not the type, but what you said makes sense. Thanks for pointing this out!